 Fix race condition when doing GATT discovery

Right now if discovery to multiple GATT clients is happening
simultaneously, onSearchComplete will be run only for device that
finishes discovery last, while it should be run for each device.

mSearchQueue is one for all discovery sessions. Instead of checking if
it's empty, we must check if there are no elements for given connId.

diff --git a/src/com/android/bluetooth/gatt/GattService.java b/src/com/android/bluetooth/gatt/GattService.java
index 010d2b6..00fc5fa 100644
--- a/src/com/android/bluetooth/gatt/GattService.java
+++ b/src/com/android/bluetooth/gatt/GattService.java
@@ -2171,7 +2171,20 @@ private void enforcePrivilegedPermission() {
     }
 
     private void continueSearch(int connId, int status) throws RemoteException {
-        if (status == 0 && !mSearchQueue.isEmpty()) {
+
+        // Search is complete when there was error, or nothing more to process
+        if (status != 0 || mSearchQueue.isEmptyFor(connId)) {
+            // In case we complete because of error, clean up
+            // any remaining operations for this connection.
+            mSearchQueue.removeConnId(connId);
+
+            ClientMap.App app = mClientMap.getByConnId(connId);
+            if (app != null) {
+                app.callback.onSearchComplete(mClientMap.addressByConnId(connId), status);
+            }
+        }
+
+        if (!mSearchQueue.isEmpty()) {
             SearchQueue.Entry svc = mSearchQueue.pop();
 
             if (svc.charUuidLsb == 0) {
@@ -2184,11 +2197,6 @@ private void continueSearch(int connId, int status) throws RemoteException {
                     svc.srvcInstId, svc.srvcUuidLsb, svc.srvcUuidMsb,
                     svc.charInstId, svc.charUuidLsb, svc.charUuidMsb, 0, 0, 0);
             }
-        } else {
-            ClientMap.App app = mClientMap.getByConnId(connId);
-            if (app != null) {
-                app.callback.onSearchComplete(mClientMap.addressByConnId(connId), status);
-            }
         }
     }
 
diff --git a/src/com/android/bluetooth/gatt/SearchQueue.java b/src/com/android/bluetooth/gatt/SearchQueue.java
index 020d2b6..00fc5fa 100644
--- a/src/com/android/bluetooth/gatt/SearchQueue.java
+++ b/src/com/android/bluetooth/gatt/SearchQueue.java
@@ -86,6 +86,16 @@ boolean isEmpty() {
         return mEntries.isEmpty();
     }
 
+    boolean isEmptyFor(int connId) {
+        for (Iterator<Entry> it = mEntries.iterator(); it.hasNext();) {
+            Entry entry = it.next();
+            if (entry.connId == connId) {
+                return false;
+            }
+        }
+        return true;
+    }
+
     void clear() {
         mEntries.clear();
     }
